mbedtls/src/ssl/context.rs

```diff
@@ -6,42 +6,38 @@
  * option. This file may not be copied, modified, or distributed except
  * according to those terms. */
 
+
+use core::any::Any;
+use core::result::Result as StdResult;
 #[cfg(not(feature = "std"))]
-use core_io::{self as io, Read, Write};
+use core_io::{Read, Write, Result as IoResult, Error as IoError, ErrorKind as IoErrorKind};
+#[cfg(feature = "std")]
+use std::io::{Read, Write, Result as IoResult, Error as IoError, ErrorKind as IoErrorKind};
 #[cfg(feature = "std")]
-use std::io::{self, Read, Write};
+use std::sync::Arc;
 
 use mbedtls_sys::types::raw_types::{c_int, c_uchar, c_void};
 use mbedtls_sys::types::size_t;
 use mbedtls_sys::*;
 
-use crate::error::{Error, IntoResult, Result};
-use core::result::Result as StdResult;
+use crate::alloc::{List as MbedtlsList};
+#[cfg(not(feature = "std"))]
+use crate::alloc_prelude::*;
+use crate::error::{Error, Result, IntoResult};
+use crate::pk::Pk;
 use crate::private::UnsafeFrom;
-use crate::ssl::config::{AuthMode, Config, Version};
-use crate::x509::{Crl, LinkedCertificate, VerifyError};
-
-pub trait IoCallback {
-    unsafe extern "C" fn call_recv(
-        user_data: *mut c_void,
-        data: *mut c_uchar,
-        len: size_t,
-    ) -> c_int;
-    unsafe extern "C" fn call_send(
-        user_data: *mut c_void,
-        data: *const c_uchar,
-        len: size_t,
-    ) -> c_int;
+use crate::ssl::config::{Config, Version, AuthMode};
+use crate::x509::{Certificate, Crl, VerifyError};
+
 
+pub trait IoCallback : Any {
+    unsafe extern "C" fn call_recv(user_data: *mut c_void, data: *mut c_uchar, len: size_t) -> c_int where Self: Sized;
+    unsafe extern "C" fn call_send(user_data: *mut c_void, data: *const c_uchar, len: size_t) -> c_int where Self: Sized;
     fn data_ptr(&mut self) -> *mut c_void;
 }
 
-impl<IO: Read + Write> IoCallback for IO {
-    unsafe extern "C" fn call_recv(
-        user_data: *mut c_void,
-        data: *mut c_uchar,
-        len: size_t,
-    ) -> c_int {
+impl<IO: Read + Write + 'static> IoCallback for IO {
```

Comment:
I'll add the 'static bound to other callbacks as well. I can't have members hold references.

Reply:
This is needed due to IoCallback deriving from Any now.

Adding this:
```
use std::any::Any;

pub trait IoCallback : Any {
```

Results in:
```
error[E0310]: the parameter type `IO` may not live long enough
  --> mbedtls/src/ssl/context.rs:41:24
   |
41 | impl<IO: Read + Write> IoCallback for IO {
   |      ---               ^^^^^^^^^^ ...so that the type `IO` will meet its required lifetime bounds
   |      |
   |      help: consider adding an explicit lifetime bound...: `IO: 'static +`
```